# InfoVis Projekt

In [1]:
import pandas as pd

1) Laden Sie Ihren Datensatz in das Notebook:

In [3]:
df = pd.read_csv("./heart_2020_cleaned.csv", sep=",")
df

HeartDisease    BMI Smoking AlcoholDrinking Stroke  PhysicalHealth  \
0                No  16.60     Yes              No     No             3.0   
1                No  20.34      No              No    Yes             0.0   
2                No  26.58     Yes              No     No            20.0   
3                No  24.21      No              No     No             0.0   
4                No  23.71      No              No     No            28.0   
...             ...    ...     ...             ...    ...             ...   
319790          Yes  27.41     Yes              No     No             7.0   
319791           No  29.84     Yes              No     No             0.0   
319792           No  24.24      No              No     No             0.0   
319793           No  32.81      No              No     No             0.0   
319794           No  46.56      No              No     No             0.0   

        MentalHealth DiffWalking     Sex  AgeCategory      Race Diabetic  \
0               30.0          No  Female        55-59     White      Yes   
1                0.0          No  Female  80 or older     White       No   
2               30.0          No    Male        65-69     White      Yes   
3                0.0          No  Female        75-79     White       No   
4                0.0         Yes  Female        40-44     White       No   
...              ...         ...     ...          ...       ...      ...   
319790           0.0         Yes    Male        60-64  Hispanic      Yes   
319791           0.0          No    Male        35-39  Hispanic       No   
319792           0.0          No  Female        45-49  Hispanic       No   
319793           0.0          No  Female        25-29  Hispanic       No   
319794           0.0          No  Female  80 or older  Hispanic       No   

       PhysicalActivity  GenHealth  SleepTime Asthma KidneyDisease SkinCancer  
0                   Yes  Very good        5.0    Yes            No        Yes  
1                   Yes  Very good        7.0     No            No         No  
2                   Yes       Fair        8.0    Yes            No         No  
3                    No       Good        6.0     No            No        Yes  
4                   Yes  Very good        8.0     No            No         No  
...                 ...        ...        ...    ...           ...        ...  
319790               No       Fair        6.0    Yes            No         No  
319791              Yes  Very good        5.0    Yes            No         No  
319792              Yes       Good        6.0     No            No         No  
319793               No       Good       12.0     No            No         No  
319794              Yes       Good        8.0     No            No         No  

[319795 rows x 18 columns]

2) Kurze Beschreibung des Datensatzes: worum geht es und woher haben Sie den Datensatz (Link)?

Link: https://www.kaggle.com/datasets/kamilpytlak/personal-key-indicators-of-heart-disease

Der Datensatz enthält Informationen zu verschiedenen persönlichen Indikatoren, die mit Herzkrankheiten in Verbindung stehen können. Die Spalten enthalten Informationen über den Body Mass Index (BMI), Rauchen, Alkoholkonsum, Schlaganfall, körperliche und geistige Gesundheit, Schwierigkeiten beim Gehen, Geschlecht, Alterskategorie, Rasse, Diabetes, körperliche Aktivität, allgemeine Gesundheit, Schlafzeit, Asthma, Nierenerkrankungen und Hautkrebs. Der Zweck des Datensatzes besteht darin, Muster und Zusammenhänge zwischen diesen Indikatoren und Herzkrankheiten zu identifizieren und zu untersuchen.


4.) Wie groß ist der Datensatz in Bezug auf Spalten und Zeilen? Welche Spalten und Zeilen sind relevant für das Projekt?

*Anmerkung: nicht benötigte Spalten können mittels `drop` entfernt werden.*

Original:   18 Spalten und 319795 Zeilen

Aus meiner Sicht besitzt jede Spalte aus diesem Datensatz potenziell wertvolle Informationen. Die Datenvisualisierung soll Auswirkungen auf eine Herzkrankheit aufzeigen, es könnte zwar sein, dass gewisse Spalten/Indikatoren keine signifikanten Auswirkungen auf eine Herzkrankheit haben, jedoch würde eine solche Analyse (welche Indikatoren Auswirkungen haben und welche nicht) den Rahmen dieser Lehrveranstaltung aus meiner Sicht sprengen.  

In [4]:
df.shape

(319795, 18)

5.) Gibt es fehlende Werte in dem Datensatz? Wenn ja: Wie gehen Sie damit um?

In [5]:
df.isnull().sum()

HeartDisease        0
BMI                 0
Smoking             0
AlcoholDrinking     0
Stroke              0
PhysicalHealth      0
MentalHealth        0
DiffWalking         0
Sex                 0
AgeCategory         0
Race                0
Diabetic            0
PhysicalActivity    0
GenHealth           0
SleepTime           0
Asthma              0
KidneyDisease       0
SkinCancer          0
dtype: int64

Laut der oberen Überprüfung ist der Datensatz vollständig. Jedoch kann so noch nicht ausgeschlossen werden, dass die erhobenen Daten auch korrekt sind. Da ich inkorrekte Daten sowieso nicht identifizieren bzw. korrigieren könnte, wird der Datensatz als vollständig und korrekt betrachtet.

---

### Data Abstraction

6.) Welche Attributtypen können den einzelnen Spalten zugewiesen werden? Geben Sie für ordinale und nominale Daten an, wieviele Kategorien pro Spalte enthalten sind! Geben Sie für quantitative Daten den Wertebereich an (Minimum und Maximum)!

In [12]:
# show data types
df.info()


<class 'pandas.core.frame.DataFrame'>
RangeIndex: 319795 entries, 0 to 319794
Data columns (total 18 columns):
 #   Column            Non-Null Count   Dtype  
---  ------            --------------   -----  
 0   HeartDisease      319795 non-null  object 
 1   BMI               319795 non-null  float64
 2   Smoking           319795 non-null  object 
 3   AlcoholDrinking   319795 non-null  object 
 4   Stroke            319795 non-null  object 
 5   PhysicalHealth    319795 non-null  float64
 6   MentalHealth      319795 non-null  float64
 7   DiffWalking       319795 non-null  object 
 8   Sex               319795 non-null  object 
 9   AgeCategory       319795 non-null  object 
 10  Race              319795 non-null  object 
 11  Diabetic          319795 non-null  object 
 12  PhysicalActivity  319795 non-null  object 
 13  GenHealth         319795 non-null  object 
 14  SleepTime         319795 non-null  float64
 15  Asthma            319795 non-null  object 
 16  KidneyDisease     31

In [26]:
# show attribute types
print("\nDie folgenden Eigenschaften des Datensatzes sind nominal: {}, {}, {}, {}, {}, {}, {}, {}, {}, {}, {} und {}."
      .format   (   df.columns[0],
                    df.columns[2],
                    df.columns[3],
                    df.columns[4],
                    df.columns[7],
                    df.columns[8],
                    df.columns[10],
                    df.columns[11],
                    df.columns[12],
                    df.columns[15],
                    df.columns[16],
                    df.columns[17]
                ))

print("Die Spalte {} enthält {} verschiedene Werte: {}.".format (   df.columns[0], 
                                                                    len(df["HeartDisease"].unique()),
                                                                    df["HeartDisease"].unique()                                                   
                                                                ))

print("Die Spalte {} enthält {} verschiedene Werte: {}.".format (   df.columns[2], 
                                                                    len(df["Smoking"].unique()),
                                                                    df["Smoking"].unique()                                                   
                                                                ))

print("Die Spalte {} enthält {} verschiedene Werte: {}.".format (   df.columns[3], 
                                                                    len(df["AlcoholDrinking"].unique()),
                                                                    df["AlcoholDrinking"].unique()                                                   
                                                                ))

print("Die Spalte {} enthält {} verschiedene Werte: {}.".format (   df.columns[4], 
                                                                    len(df["Stroke"].unique()),
                                                                    df["Stroke"].unique()                                                   
                                                                ))

print("Die Spalte {} enthält {} verschiedene Werte: {}.".format (   df.columns[7], 
                                                                    len(df["DiffWalking"].unique()),
                                                                    df["DiffWalking"].unique()                                                   
                                                                ))

print("Die Spalte {} enthält {} verschiedene Werte: {}.".format (   df.columns[8], 
                                                                    len(df["Sex"].unique()),
                                                                    df["Sex"].unique()                                                   
                                                                ))

print("Die Spalte {} enthält {} verschiedene Werte: {}.".format (   df.columns[10], 
                                                                    len(df["Race"].unique()),
                                                                    df["Race"].unique()                                                   
                                                                ))

print("Die Spalte {} enthält {} verschiedene Werte: {}.".format (   df.columns[11], 
                                                                    len(df["Diabetic"].unique()),
                                                                    df["Diabetic"].unique()                                                   
                                                                ))

print("Die Spalte {} enthält {} verschiedene Werte: {}.".format (   df.columns[12], 
                                                                    len(df["PhysicalActivity"].unique()),
                                                                    df["PhysicalActivity"].unique()                                                   
                                                                ))

print("Die Spalte {} enthält {} verschiedene Werte: {}.".format (   df.columns[15], 
                                                                    len(df["Asthma"].unique()),
                                                                    df["Asthma"].unique()                                                   
                                                                ))

print("Die Spalte {} enthält {} verschiedene Werte: {}.".format (   df.columns[16], 
                                                                    len(df["KidneyDisease"].unique()),
                                                                    df["KidneyDisease"].unique()                                                   
                                                                ))

print("Die Spalte {} enthält {} verschiedene Werte: {}.".format (   df.columns[17], 
                                                                    len(df["SkinCancer"].unique()),
                                                                    df["SkinCancer"].unique()                                                   
                                                                ))


Die folgenden Eigenschaften des Datensatzes sind nominal: HeartDisease, Smoking, AlcoholDrinking, Stroke, DiffWalking, Sex, Race, Diabetic, PhysicalActivity, Asthma, KidneyDisease und SkinCancer.
Die Spalte HeartDisease enthält 2 verschiedene Werte: ['No' 'Yes'].
Die Spalte Smoking enthält 2 verschiedene Werte: ['Yes' 'No'].
Die Spalte AlcoholDrinking enthält 2 verschiedene Werte: ['No' 'Yes'].
Die Spalte Stroke enthält 2 verschiedene Werte: ['No' 'Yes'].
Die Spalte DiffWalking enthält 2 verschiedene Werte: ['No' 'Yes'].
Die Spalte Sex enthält 2 verschiedene Werte: ['Female' 'Male'].
Die Spalte Race enthält 6 verschiedene Werte: ['White' 'Black' 'Asian' 'American Indian/Alaskan Native' 'Other'
 'Hispanic'].
Die Spalte Diabetic enthält 4 verschiedene Werte: ['Yes' 'No' 'No, borderline diabetes' 'Yes (during pregnancy)'].
Die Spalte PhysicalActivity enthält 2 verschiedene Werte: ['Yes' 'No'].
Die Spalte Asthma enthält 2 verschiedene Werte: ['Yes' 'No'].
Die Spalte KidneyDisease enthält 

In [27]:
print("\nDie folgenden Eigenschaften des Datensatzes sind ordinal: {} und {}."
      .format   (   df.columns[9],
                    df.columns[13]
                ))

print("Die Spalte {} enthält {} verschiedene Werte: {}.".format (   df.columns[9], 
                                                                    len(df["AgeCategory"].unique()),
                                                                    df["AgeCategory"].unique()                                                   
                                                                ))

print("Die Spalte {} enthält {} verschiedene Werte: {}.".format (   df.columns[13], 
                                                                    len(df["GenHealth"].unique()),
                                                                    df["GenHealth"].unique()    
                                                                ))


Die folgenden Eigenschaften des Datensatzes sind ordinal: AgeCategory und GenHealth.
Die Spalte AgeCategory enthält 13 verschiedene Werte: ['55-59' '80 or older' '65-69' '75-79' '40-44' '70-74' '60-64' '50-54'
 '45-49' '18-24' '35-39' '30-34' '25-29'].
Die Spalte GenHealth enthält 5 verschiedene Werte: ['Very good' 'Fair' 'Good' 'Poor' 'Excellent'].


In [28]:
print("\nDie folgenden Eigenschaften des Datensatzes sind quantitative: {}, {}, {} und {}."
      .format   (   df.columns[1],
                    df.columns[5],
                    df.columns[6],
                    df.columns[14]
                ))

print("Der Wertebereich der Spalte {} reicht von {} bis {}.\n".format   (   df.columns[1], 
                                                                            df["BMI"].min(),
                                                                            df["BMI"].max()    
                                                                        ))

print("Der Wertebereich der Spalte {} reicht von {} bis {}.\n".format   (   df.columns[5], 
                                                                            df["PhysicalHealth"].min(),
                                                                            df["PhysicalHealth"].max()    
                                                                        ))

print("Der Wertebereich der Spalte {} reicht von {} bis {}.\n".format   (   df.columns[6], 
                                                                            df["MentalHealth"].min(),
                                                                            df["MentalHealth"].max()    
                                                                        ))

print("Der Wertebereich der Spalte {} reicht von {} bis {}.\n".format   (   df.columns[14], 
                                                                            df["SleepTime"].min(),
                                                                            df["SleepTime"].max()    
                                                                        ))


Die folgenden Eigenschaften des Datensatzes sind quantitative: BMI, PhysicalHealth, MentalHealth und SleepTime.
Der Wertebereich der Spalte BMI reicht von 12.02 bis 94.85.

Der Wertebereich der Spalte PhysicalHealth reicht von 0.0 bis 30.0.

Der Wertebereich der Spalte MentalHealth reicht von 0.0 bis 30.0.

Der Wertebereich der Spalte SleepTime reicht von 1.0 bis 24.0.



7.) Welche Datenstruktur(en) stecken in dem Datensatz? (z.B. hierarchisch, geografisch, temporal)

Der Datensatz ist eine flache (nicht hierarchische) und zeitunabhängige Datenstruktur. Es gibt keine Hierarchie zwischen den Daten und keine geografischen Informationen, die Informationen beinhalten würden, wo die Daten gesammelt wurden. Außerdem gibt es keine zeitbezogenen Informationen wie Datum oder Uhrzeit, die Aussagen, wann die Daten gesammelt wurden.

### Task Abstraction

8.) Welche Fragen können mit dem gewählten Datensatz untersucht werden?

*Anmerkung: Diese Fragen können in einem domänenspezifischen Format gestellt werden. Hinterfragen Sie hier auch: welche Fragen benötigen eine visuelle Analyse? Nicht nur Bestimmung von Anzahl und Min/Max*

- Wie hängen körperliche Aktivitäten und körperliche Gesundheit mit dem Risiko von Herzerkrankungen zusammen?
- Welche Auswirkungen hat das Rauchen auf das Risiko von Herzerkrankungen im Vergleich zum Alkoholkonsum?
- Welchen Einfluss hat das Alter auf das Risiko von Herzerkrankungen?
- Gibt es einen Zusammenhang zwischen der genetischen Gesundheit und dem Risiko von Herzerkrankungen?
- Wie beeinflussen bestimmte ethnische Gruppen das Risiko von Herzerkrankungen?
- Welche Rolle spielen Schlafstörungen bei der Entwicklung von Herzerkrankungen?
- Wie beeinflusst das Vorhandensein von Diabetes, Asthma oder Nierenerkrankungen das Risiko von Herzerkrankungen?
- Wie wirkt sich die mentale Gesundheit auf das Risiko von Herzerkrankungen aus?

9.) Welche abstrakten Aufgaben stecken in den Fragen? *(z.B. Finden von Ausreißern, Vergleich von mehreren Werten)*

- Bewertung von Risikofaktoren und deren Beitrag zum Gesamtrisiko von Herzerkrankungen
- Vergleich von mehreren Messungen oder Indikatoren, um das Risiko von Herzerkrankungen vorherzusagen oder zu verstehen
- Identifizierung von Ausreißern oder Abweichung von erwarteten Mustern
- Vergleich von Gruppen und Subgruppen anhand verschiedener Variablen
- Identifizierung von Korrelationen und Zusammenhängen zwischen Variablen
- Untersuchung von Trends in verschiedenen Altergruppen
- Analyse der Verteilung von Daten und der Häufigkeit von bestimmten Merkmalen

10.) Welche Datentransformationen sind nötig um die Fragen zu beantworten? 
*(z.B. Aggregation, Mergen von zwei Datensätzen, Umwandlung von Attributtypen)*

- Aggregation von Daten auf verschiedenen Ebenen, z.B. durch Berechnung von Durchschnittswerten oder Summen pro Gruppe (z.B. nach Altergruppe)
- Filterung von Daten, um den Fokus auf bestimmte Subgruppen zu legen (z.B. Raucher, Personen mit hohem BMI)
- Transformation von Attributtypen, um sie für bestimmte Analysemethoden oder Visualisierungen zu verwenden
- Skalierung von Variablen auf einen anderen Wertebereich
- Berechnung von Durchschnittswerten, bilden von Minima bzw. Maxima
